src/EditorFeatures/CSharpTest/CodeActions/ConvertLinq/ConvertLinqQueryToForEachTests.cs

```diff
@@ -0,0 +1,2246 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.CodeRefactorings;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.CodeActions.ConvertLinq
+{
+    public class ConvertLinqQueryToForEachTests : AbstractCSharpCodeActionTest
+    {
+     
+        protected override CodeRefactoringProvider CreateCodeRefactoringProvider(Workspace workspace, TestParameters parameters)
+          => new CodeAnalysis.CSharp.ConvertLinq.CSharpConvertLinqQueryToLinqMethodProvider();
+
+        // TODO test case within array initialization
+        // TODO test case within a list/struct initializator
+        // TODO test case within a ctor
+        // TODO add test for from x in XX where XX is not IEnumerable, i.e. requires AsEnumerable.
+        // TODO unit test with mutliple selects
+
+        [Fact]
+        public async Task AnonymousType()
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    bool M(IEnumerable<int> nums)
+    {
+        var q = [|from a in nums from b in nums select new { a, b }|];
+    }
+}
+";
+            // No conversion can be made because it expects to introduce a local funciton but the return type contains anonymous.
+            await TestMissingAsync(source);
+        }
+
+        [Fact]
+        public async Task ReturnIEnumerable()
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    IEnumerable<int> M(IEnumerable<int> nums)
+    {
+        return [|from int n1 in nums 
+                 from int n2 in nums
+                 select n1|];
+    }
+}
+";
+
+            string output = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    IEnumerable<int> M(IEnumerable<int> nums)
+    {
+        foreach (int n1 in nums)
+        {
+            foreach (int n2 in nums)
+            {
+                yield return n1;
+            }
+        }
+    }
+}
+";
+
+            await TestInRegularAndScriptAsync(source, output);
+        }
+
+        [Fact]
+        public async Task MultipleAssignments()
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Linq;
+public class Test
+{
+    public static void Main()
+    {
+        var nums = new int[] { 1, 2, 3, 4 };
+        IEnumerable<int> q1, q2;
+        q1 = q2 = [|from x in nums select x + 1|];
+    }
+}";
+
+            string output = @"
+using System.Collections.Generic;
+using System.Linq;
+public class Test
+{
+    public static void Main()
+    {
+        var nums = new int[] { 1, 2, 3, 4 };
+        IEnumerable<int> q1, q2;
+        IEnumerable<int> localFunction()
```

Comment:
>IEnumerable<int> [](start = 8, length = 16)

Do we have a test where `using System.Collections.Generic` is not present to confirm that we fully qualify the return type?

Reply:
sure, e.g. please look for RangeVariables

---
In reply to: [176553315](https://github.com/dotnet/roslyn/pull/25362#discussion_r176553315) [](ancestors = 176553315)